race-buffer/src/async_reader.rs

```diff
@@ -0,0 +1,286 @@
+            n_missed_before_read += 1;
+        }
+
+        // Store valid entries in read buffer
+        for entry in &mut buf_snapshot[Into::<u64>::into(n_overwritten_in_snap) as usize..] {
+            self.store(*entry, rbuf);
+        }
+        Ok((n_missed_before_read + n_overwritten_in_snap).into())
+    }
+
+    /// Store given entry in given read buffer
+    #[inline]
+    fn store(&mut self, entry: E, rbuf: &mut Vec<WholeEntry<E>>) {
+        if let Some(prefix) = self.stored_prefix.take() {
+            rbuf.push(WholeEntry::Double(prefix, entry));
+        } else if entry.is_prefix() {
+            self.stored_prefix = Some(entry);
+        } else {
+            rbuf.push(WholeEntry::Single(entry));
+        }
+    }
+
+    /// Drop stored prefix, returning false if it was already None
+    #[inline]
+    fn drop_prefix(&mut self) -> bool {
+        self.stored_prefix.take().is_some()
+    }
+
+    fn snap_storage(&self, index: usize) -> Result<E, S::Error> {
+        self.snapper.snap_storage(index)
+    }
+
+    fn snap_write_seqn(&mut self) -> Result<SeqNum, S::Error> {
+        Self::snap_seqn(
+            || self.snapper.snap_write_seqn_high(),
+            || self.snapper.snap_write_seqn_low(),
+        )
+    }
+
+    fn snap_overwrite_seqn(&mut self) -> Result<SeqNum, S::Error> {
+        Self::snap_seqn(
+            || self.snapper.snap_overwrite_seqn_high(),
+            || self.snapper.snap_overwrite_seqn_low(),
+        )
+    }
+
+    /// Get a consistent snapshot of a sequence number, retrying if inconsistency is detected
+    fn snap_seqn<F, G>(snap_high: F, snap_low: G) -> Result<SeqNum, S::Error>
+    where
+        F: Fn() -> Result<u32, S::Error>,
+        G: Fn() -> Result<u32, S::Error>,
+    {
+        let mut initial_high;
+        let mut final_high;
+        let mut low;
+        // Note: This will loop until a consistent snapshot is acquired. If the high word is incremented during
+        // the snapshot, then the snapshot will not be consistent. However, the high word is only incremented every
+        // 2^32 entries written to the buffer. We assume that the amount of time it takes to take a snapshot is
+        // negligible compared to the time it takes to increment the high word. Therefore, there should only be one
+        // retry at most, in the case where the first snapshot attempt and an increment to the high word happen to overlap.
+        while {
+            // Wait until the sequence number high word is not getting updated
+            initial_high = Self::snap_consistent_high(&snap_high)?;
+            low = snap_low()?;
+            final_high = snap_high()?;
+            // Check that the high word did not change between the two snapshots. If it did,
+            // The low word is unusable because we are not sure which high word it corresponds to.
+            // In that case, retry the whole sequence
+            initial_high != final_high
```

Comment:
I have to say, this is one of the cooler uses of "blocks are expressions" in Rust.